### PR TITLE
Add HDFS exporter

### PR DESCRIPTION
**Pull request by gray-matter**
Take events from Kafka and push them to HDFS as Protobuf + Parquet

### CODE REVIEW

File: readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriter.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+        @Override
+        public void run() {
+            runningThread = Thread.currentThread();
+
+            while (!runningThread.isInterrupted()) {
+                writers.forEach(PartitionedWriter::expireConsumers);
+
+                try {
+                    Thread.sleep(period.get(ChronoUnit.SECONDS) * 1000);
```
[the remark this answers is not in the document]
Comment (gray-matter) [reply readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriter.java]: > Exception in thread "Thread-0" java.time.temporal.UnsupportedTemporalTypeException: Unsupported unit: Millis

:wink:

File: readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriter.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+    /**
+     * Poll a list of PartitionedWriter instances to make them expire if relevant. To be executed in a separate thread.
+     *
+     * @param <MessageKind> Type of messages which will ultimately get written.
+     */
+    public static class Expirer<MessageKind> implements Runnable {
+        private final Collection<PartitionedWriter<MessageKind>> writers;
+        private final TemporalAmount period;
+        private Thread runningThread;
```
[the remark this answers is not in the document]
Comment (gray-matter) [reply readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriter.java]: Done

File: readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriter.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+     * @param partitionId       Origin partition id
+     * @return                  Existing or just-created consumer
+     */
+    private ExpiringConsumer<MessageKind> getWriter(LocalDateTime dayStartTime, int partitionId) {
+        if (!perPartitionDayWriters.containsKey(partitionId))
+            perPartitionDayWriters.put(partitionId, new HashMap<>());
+
+        final Map<LocalDateTime, ExpiringConsumer<MessageKind>> partitionMap = perPartitionDayWriters.get(partitionId);
+
+        if (!partitionMap.containsKey(dayStartTime))
```
[the remark this answers is not in the document]
Comment (gray-matter) [reply readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriter.java]: Done

File: readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputer.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        while (filesIterator.hasNext()) {
+            final LocatedFileStatus fileStatus = filesIterator.next();
+            final String fileName = fileStatus.getPath().getName();
+
+            final Matcher matcher = partitionBasedPattern.matcher(fileName);
+            if (!matcher.matches() || matcher.groupCount() < 1)
+                continue;
+
+            final String offsetString = matcher.group(1);
```
[the remark this answers is not in the document]
Comment (gray-matter) [reply readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputer.java]: Agree, I circumvented the issue by creating an integration test, but it's still a bit cumbersome. I will try merging FileNamer and OffsetComputer

File: readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/HdfsExporter.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        final KafkaConsumer<String, byte[]> kafkaConsumer = new KafkaConsumer<>(props);
+        final GarmadonReader.Builder readerBuilder = GarmadonReader.Builder.stream(kafkaConsumer);
+        final Collection<PartitionedWriter<Message>> writers = new ArrayList<>();
+        final PartitionedWriter.Expirer expirer = new PartitionedWriter.Expirer<>(writers, Duration.ofSeconds(10));
+        final Thread expirerThread = new Thread(expirer);
+        final HeartbeatConsumer heartbeat = new HeartbeatConsumer<>(writers, Duration.ofSeconds(5));
+        final Thread heartbeatThread = new Thread(heartbeat);
+        final Map<Integer, Map.Entry<String, Class<? extends Message>>> typeToDirAndClass = getTypeToDirAndClass();
+        final Random random = buildMachineUniqueRandom();
+        final Path temporaryHdfsDir = new Path(baseTemporaryHdfsDir, String.valueOf(random.nextInt(Integer.MAX_VALUE)));
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/HdfsExporter.java]: agreed with a simple random UUID

File: readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriter.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    /**
+     * Get the starting offset for a given partition.
+     *
+     * @param partitionId
+     * @return              The lowest offset for a given partition
+     * @throws IOException  If the offset computation failed
+     */
+    public long getStartingOffset(int partitionId) throws IOException {
+        synchronized (perPartitionStartOffset) {
+            if (!perPartitionStartOffset.containsKey(partitionId)) {
```
[the remark this answers is not in the document]
Comment (Willymontaz) [reply readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriter.java]: I did not see that, so ok for me

File: readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/HdfsExporter.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        };
+    }
+
+    private static GarmadonReader.GarmadonMessageHandler buildGarmadonMessageHandler(PartitionedWriter<Message> writer) {
+        return (msg) -> {
+            final CommittableOffset offset = msg.getCommittableOffset();
+
+            try {
+                writer.write(Instant.now(), offset,
+                        ProtoConcatenator.concatToProtobuf(Arrays.asList(msg.getHeader(), (Message) msg.getBody())));
```
[the remark this answers is not in the document]
Comment (gray-matter) [reply readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/HdfsExporter.java]: As discussed together, we'll use a simple cache in a future version if this becomes an issue. v1 will not include it

**Pull request merged**